### PR TITLE
arch/risc-v: move REGLOAD/REGSTORE macro to riscv_internal.h

### PR DESCRIPTION
**Pull request by pkarashchenko**
## Summary
Improve RISC-V common code usage

## Impact
Should not be any. Refactoring only

## Testing
Pass CI

### CODE REVIEW

File: arch/risc-v/src/common/riscv_swint.c [hunk 1]
```diff
@@ -109,45 +109,24 @@ static void riscv_registerdump(const uintptr_t *regs)
...
-     " slli a0, a0, 3\n"          /* a0=Offset for the stub lookup table */
-     " add  t0, t0, a0\n"         /* t0=The address in the table */
-     " ld   t0, 0(t0)\n"          /* t0=The address of the stub for this syscall */
-     " jalr ra, t0\n"             /* Call the stub (modifies ra) */
-     " ld   ra, 0(sp)\n"          /* Restore ra */
-     " addi sp, sp, 8\n"          /* Destroy the stack frame */
-     " mv   a2, a0\n"             /* a2=Save return value in a0 */
-     " li   a0, 3\n"              /* a0=SYS_syscall_return (3) */
-     " ecall"                     /* Return from the syscall */
+     "addi sp, sp, -" REGSSIZE "\n" /* Create a stack frame to hold ra */
```
Comment (pussuw) [inline arch/risc-v/src/common/riscv_swint.c:116]: Using the register size here is incorrect. The stack pointer has separate alignment requirements, and in this case a 16B alignment should be ensured, if FPU is in use, if I remember correctly. 

It is IMO simpler to just make a 16B stack frame here. Every other place I could find does this, and this is what I did for the S-mode patch too. There is a macro for this too in riscv_internal.h `STACK_ALIGNMENT`

Comment (pkarashchenko) [reply arch/risc-v/src/common/riscv_swint.c]: Done. Please take a look

File: arch/risc-v/src/common/riscv_swint.c [hunk 2]
```diff
@@ -109,45 +109,24 @@ static void riscv_registerdump(const uintptr_t *regs)
...
+     REGSTORE " ra, 0(sp)\n"        /* Save ra in the stack frame */
+     "la   t0, g_stublookup\n"      /* t0=The base of the stub lookup table */
+     "slli a0, a0, 3\n"             /* a0=Offset for the stub lookup table */
```
[the remark this answers is not in the document]
Comment (pussuw) [reply arch/risc-v/src/common/riscv_swint.c]: I don't know for sure but yes I would assume that the original implementation is incorrect for the shift. Do any 32-bit targets use BUILD_PROTECTED ? If the element in g_stublookup is machine word aligned, then the shift should be log2() of word length.

[the remark this answers is not in the document]
Comment (pkarashchenko) [reply arch/risc-v/src/common/riscv_swint.c]: Now I need to think about good macro for that :)

[the remark this answers is not in the document]
Comment (pkarashchenko) [reply arch/risc-v/src/common/riscv_swint.c]: @masayuki2009 @xiaoxiang781216 please still take a look if the applied change is correct. I do not have possibility to test it.

File: arch/risc-v/src/common/riscv_signal_handler.S [hunk 3]
```diff
@@ -69,7 +71,7 @@ up_signal_handler:
 
   /* Save ra on the stack */
 
-  addi sp, sp, -16
+  addi sp, sp, -STACK_FRAME_SIZE
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_signal_handler.S:74]: should we use STACK_ALIGNMENT here?

File: arch/risc-v/src/common/riscv_signal_handler.S [hunk 4]
```diff
@@ -83,7 +85,7 @@ up_signal_handler:
   /* Restore the register */
 
   ld   ra, 8(sp)  /* Restore ra in sp */
-  addi sp, sp, 16
+  addi sp, sp, STACK_FRAME_SIZE
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_signal_handler.S:88]: ditto

File: arch/risc-v/src/common/riscv_swint.c [hunk 5]
```diff
@@ -109,45 +109,24 @@ static void riscv_registerdump(const uintptr_t *regs)
...
-     " jalr ra, t0\n"             /* Call the stub (modifies ra) */
-     " ld   ra, 0(sp)\n"          /* Restore ra */
-     " addi sp, sp, 8\n"          /* Destroy the stack frame */
-     " mv   a2, a0\n"             /* a2=Save return value in a0 */
-     " li   a0, 3\n"              /* a0=SYS_syscall_return (3) */
-     " ecall"                     /* Return from the syscall */
+     "addi sp, sp, -" STACK_FRAME_SIZE "\n" /* Create a stack frame to hold ra */
+     REGSTORE " ra, 0(sp)\n"                /* Save ra in the stack frame */
+     "la   t0, g_stublookup\n"              /* t0=The base of the stub lookup table */
+     "slli a0, a0, 3\n"                     /* a0=Offset for the stub lookup table */
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_swint.c:122]: Should we use the different shift here(2 for 32bit and 3 for 64bit)?

File: arch/risc-v/src/common/riscv_signal_handler.S (hunk 3 above)
Comment (pkarashchenko) [reply arch/risc-v/src/common/riscv_signal_handler.S:74]: Yes. The STACK_ALIGNMENT value is used

File: arch/risc-v/src/common/riscv_swint.c (hunk 5 above)
Comment (pkarashchenko) [reply arch/risc-v/src/common/riscv_swint.c:122]: That is exactly the question that I asked. Seems like we should

File: arch/risc-v/src/common/riscv_signal_handler.S (hunk 3 above)
Comment (xiaoxiang781216) [reply arch/risc-v/src/common/riscv_signal_handler.S:74]: I mean we don't need quote version here.

Comment (pkarashchenko) [reply arch/risc-v/src/common/riscv_signal_handler.S:74]: I did that just to keep same style as in dispatch_syscall, but I can change it.

Comment (pkarashchenko) [reply arch/risc-v/src/common/riscv_signal_handler.S:74]: The assembly version of STACK_FRAME_SIZE is not quoted

Comment (xiaoxiang781216) [reply arch/risc-v/src/common/riscv_signal_handler.S:74]: Ok, got it.

File: arch/risc-v/src/common/riscv_signal_handler.S [hunk 6]
```diff
@@ -83,7 +85,7 @@ up_signal_handler:
   /* Restore the register */
 
   ld   ra, 8(sp)  /* Restore ra in sp */
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_signal_handler.S:87]: Probably this place also should be modified to use REGLOAD

Comment (pkarashchenko) [reply arch/risc-v/src/common/riscv_signal_handler.S]: Fixed

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/risc-v/src/common/riscv_swint.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/risc-v/src/common/riscv_swint.c]: let's fix this issue in another patch

[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/risc-v/src/common/riscv_swint.c]: Look like you fix it.:)

**Comment by pussuw** [comment]
There is a build failure:

 In file included from common/riscv_pmp.c:34:
common/riscv_pmp.c: In function 'pmp_read_region_cfg':
Error: common/riscv_internal.h:155:20: error: declaration of 'tmp' shadows a previous local [-Werror=shadow]
      unsigned long tmp; \
                    ^~~
common/riscv_pmp.c:74:21: note: in expansion of macro 'READ_CSR'
     uintptr_t tmp = READ_CSR(reg); \
                     ^~~~~~~~
common/riscv_pmp.c:233:16: note: in expansion of macro 'PMP_READ_REGION_FROM_REG'
         return PMP_READ_REGION_FROM_REG(region, pmpcfg0);
                ^~~~~~~~~~~~~~~~~~~~~~~~
common/riscv_pmp.c:74:15: note: shadowed declaration is here
     uintptr_t tmp = READ_CSR(reg); \
               ^~~
common/riscv_pmp.c:233:16: note: in expansion of macro 'PMP_READ_REGION_FROM_REG'
         return PMP_READ_REGION_FROM_REG(region, pmpcfg0);

Please, address this failure prior to merge.

**Comment by pkarashchenko** [comment]
I will fix the build in 1h

File: arch/risc-v/src/common/riscv_internal.h [hunk 7]
```diff
@@ -121,6 +150,37 @@
 #define getreg64(a)         (*(volatile uint64_t *)(a))
 #define putreg64(v,a)       (*(volatile uint64_t *)(a) = (v))
 
+#define READ_CSR(reg) \
+  ({ \
+     uintptr_t reg##_val; \
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_internal.h:155]: @pussuw build fixed here by `reg##_val`. Pleas review. I changed from `long` to `uintptr_t` and expect that it should be fine,

Comment (pussuw) [reply arch/risc-v/src/common/riscv_internal.h:155]: Yes this should produce an unique enough tmp variable. I have no idea why the compiler has not complained about this before. Seems like enough obfuscation in pre-processing will silence it (I have seen similar behavior before).

Using uintptr_t here is fine, as all actual CSRs are XLEN

**Comment by pussuw** [comment]
LGTM, let's wait for CI and then please merge this ASAP so I can resolve the upcoming merge conflicts. Thanks!

**Comment by pussuw** [comment]
Maybe my inexperience with GH review process, but should I mark the files manually as "reviewed" or can that be done "as a bunch", I press the approve button but it does not show in the approval list ?

**Comment by pkarashchenko** [comment]
> Maybe my inexperience with GH review process, but should I mark the files manually as "reviewed" or can that be done "as a bunch", I press the approve button but it does not show in the approval list ?

The overall approve is enough and I see your approval in approval list.

File: arch/risc-v/src/common/riscv_internal.h (hunk 7 above)
Comment (pkarashchenko) [reply arch/risc-v/src/common/riscv_internal.h:155]: If needed we can go even more unique with
```
#define __tmpname(name) tmp##name
#define __uname(name)   __tmpname(name)

uintptr_t __uname(__LINE__);
```

**Comment by pkarashchenko** [comment]
@masayuki2009 San, could you please review and merge?

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**